src/components/editor/Bindings/SchemaInference/Dialog.tsx

```diff
@@ -0,0 +1,415 @@
+import AlertBox from 'components/shared/AlertBox';
+import {
+    glassBkgWithoutBlur,
+    monacoEditorHeaderBackground,
+    monacoEditorWidgetBackground,
+    secondaryButtonBackground,
+    secondaryButtonHoverBackground,
+} from 'context/Theme';
+import { isEmpty, isEqual } from 'lodash';
+import { Dispatch, SetStateAction, useState } from 'react';
+import { FormattedMessage } from 'react-intl';
+import { useEffectOnce, useLocalStorage } from 'react-use';
+import getInferredSchema, {
+    InferSchemaResponse,
+} from 'services/schema-inference';
+import { stringifyJSON } from 'services/stringify';
+import { CallSupabaseResponse } from 'services/supabase';
+import { useResourceConfig_currentCollection } from 'stores/ResourceConfig/hooks';
+import { Schema } from 'types';
+import {
+    getStoredGatewayAuthConfig,
+    LocalStorageKeys,
+} from 'utils/localStorage-utils';
+
+interface Props {
+    collectionData: CollectionData;
+    open: boolean;
+    setOpen: Dispatch<SetStateAction<boolean>>;
+    height?: number;
+}
+
+const CustomWidthTooltip = styled(({ className, ...props }: TooltipProps) => (
+    <Tooltip {...props} classes={{ popper: className }} />
+))({
+    [`& .${tooltipClasses.tooltip}`]: {
+        maxWidth: 400,
+    },
+    [`& .${tooltipClasses.popper}`]: {
+        overflowWrap: 'break-word',
+    },
+});
+
+const handleInferredSchemaSuccess = (
+    response: InferSchemaResponse,
+    collectionSchema: Schema,
+    setInferredSchema: Dispatch<SetStateAction<Schema | null | undefined>>,
+    setLoading: Dispatch<SetStateAction<boolean>>
+) => {
+    setInferredSchema(
+        !isEmpty(response.schema)
+            ? { ...collectionSchema, schema: response.schema }
+            : null
+    );
+
+    setLoading(false);
+};
+
+const handleInferredSchemaFailure = (
+    error: any,
+    setInferredSchema: Dispatch<SetStateAction<Schema | null | undefined>>,
+    setLoading: Dispatch<SetStateAction<boolean>>
+) => {
+    setInferredSchema(error?.code === 404 ? null : undefined);
+
+    setLoading(false);
+};
+
+const processDraftSpecResponse = (
+    draftSpecResponse: CallSupabaseResponse<any>,
+    schemaUpdateErrored: boolean,
+    setSchemaUpdateErrored: Dispatch<SetStateAction<boolean>>,
+    setLoading: Dispatch<SetStateAction<boolean>>,
+    setCollectionData: BindingsEditorState['setCollectionData'],
+    setOpen: Dispatch<SetStateAction<boolean>>
+) => {
+    if (draftSpecResponse.error) {
+        setSchemaUpdateErrored(true);
+        setLoading(false);
+    } else if (draftSpecResponse.data && draftSpecResponse.data.length > 0) {
+        if (schemaUpdateErrored) {
+            setSchemaUpdateErrored(false);
+        }
+
+        setCollectionData({
+            spec: draftSpecResponse.data[0].spec,
+            belongsToDraft: true,
+        });
+
+        setLoading(false);
+        setOpen(false);
+    } else {
+        setSchemaUpdateErrored(true);
+        setLoading(false);
+    }
```

Comment:
Given this is a big handler I feel like the `setLoading` call should be outside of the branches to make sure it is always there (in case we add more branches)